### DISCUSSION TITLE
Calculate Solar string Temperature

### DISCUSSION THREAD

**Discussion by ER7788** [slipx06/sunsynk-power-flow-card#570] [Ideas]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

**It would be cool to calculate the Solar string temperature.**
Here is a recept for it. I am at the moment not good in programming for HA.
That is not my usual business.

Thanks for the good and enthusiastic work you have done til now.  

**Needed Parameter:** 
PV-Modul: (all on the label of the panel)
- MPP-Ref-Voltage; Vmpp (AM1.5; 25°C Vmpp=32.01V
- MPP-Ref-Temperature; (~25°C)
- Temp-coefizient_Voltage Tk -dU/°C ; (datasheet Uoc~ -0,25%/°C)
  (can also be user as a default value; Si_cells have a similar Tk)
  (For reference Umpp should be used)
- No. of PV modules in the string. No_modul
- Serial resistance of the Wires: Rser~20-200mOhm (40mOhm preset) (in the formula in Ohm)
- The calculation is valid for ~5% Isa up to ~95% Converter current capability. (Isa_min; Isa_max, valid)

Due to the fact that MPP (Maximum Power Point) is found by varying the String Voltage a mean of the last 5-10 measurements is useful. -> T_new=(Told*9+T_meas)/10
The Calculation is only valid, if all Panels of the string have the same direction and no MPP-Power optimizer is used.

**Formula:** 
Valid, if Isa_min < Isa < Isa_max
Temp_PVstring = MPP_Ref_Temp + ((Vsa+Isa*Rser)-(VmppRef*No_molul)/(No_modul) / (Vmpp*Tk_Modul*100%)

**Some Explanaition**
In my Case the Converter is producing all the possible power available from the sun except the current limitation of the converter is active at ~13.5/13.9A. I that case the solar voltage increase and the temperature measurement in invalid.
The other invalid case is too low current for example <0,5A where the MPP-function is not properly working. 

I made some calculations for Satellite PV and the results are quite good, as far as the converter is in MPP.
There may be an offset in the real temperature, which can be adjusted with the T_ref of the module. 
Values up to 65°C in summer are possible.

I am using a Growatt SPH-10kTL3 with battery ~10kWh; PV: 11x420W/10*420W. 

Thanks in advance for implementing this as a new feature.

### Expected behaviour

The result of the calculation is the PV_String Temperature by mean.

### Mode

lite

### Context / Reason

When a Solar cell is used in MPP-Mode, it is possible to calculate the actual cell temperature.
The calculation is above.

**Comment by slipx06** [comment]
Moving this to an ideas discussion.
